raft/raft.go

```diff
@@ -0,0 +1,403 @@
+		}
+
+		time.Sleep(leadershipCheckInterval)
+	}
+}
+
+// AddPeer adds a new peer to the Raft cluster.
+func (n *Node) AddPeer(peerID, peerAddr string) error {
+	if err := n.raft.AddVoter(raft.ServerID(peerID), raft.ServerAddress(peerAddr), 0, 0).Error(); err != nil {
+		return fmt.Errorf("failed to add voter: %w", err)
+	}
+	return nil
+}
+
+// RemovePeer removes a peer from the Raft cluster.
+func (n *Node) RemovePeer(peerID string) error {
+	if err := n.raft.RemoveServer(raft.ServerID(peerID), 0, 0).Error(); err != nil {
+		return fmt.Errorf("failed to remove server: %w", err)
+	}
+	return nil
+}
+
+// Apply is the public method that handles forwarding if necessary.
+func (n *Node) Apply(data []byte, timeout time.Duration) error {
+	if n.raft.State() != raft.Leader {
+		return n.forwardToLeader(data, timeout)
+	}
+	return n.applyInternal(data, timeout)
+}
+
+// applyInternal is the internal method that actually applies the data.
+func (n *Node) applyInternal(data []byte, timeout time.Duration) error {
+	future := n.raft.Apply(data, timeout)
+	if err := future.Error(); err != nil {
+		return fmt.Errorf("failed to apply log entry: %w", err)
+	}
+	return nil
+}
+
+// forwardToLeader forwards a request to the current Raft leader node. It first identifies
+// the leader's gRPC address by matching the leader ID with the peer list. Then it establishes
+// a gRPC connection to forward the request. The method handles timeouts and returns any errors
+// that occur during forwarding.
+func (n *Node) forwardToLeader(data []byte, timeout time.Duration) error {
+	leaderAddr, leaderID := n.raft.LeaderWithID()
+	if leaderID == "" {
+		return errors.New("no leader available")
+	}
+
+	n.Logger.Debug().
+		Str("leader_id", string(leaderID)).
+		Str("leader_addr", string(leaderAddr)).
+		Msg("forwarding request to leader")
+
+	var leaderGrpcAddr string
+	for _, peer := range n.Peers {
+		if raft.ServerID(peer.ID) == leaderID {
+			leaderGrpcAddr = peer.GRPCAddress
+			break
+		}
+	}
+	// Get the RPC client for the leader
+	client, err := n.rpcClient.getClient(leaderGrpcAddr)
+	if err != nil {
+		return fmt.Errorf("failed to get client for leader: %w", err)
+	}
+
+	// Create context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	// Forward the request
+	resp, err := client.ForwardApply(ctx, &pb.ApplyRequest{
+		Data:      data,
+		TimeoutMs: timeout.Milliseconds(),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to forward request: %w", err)
+	}
+
+	if !resp.GetSuccess() {
+		return fmt.Errorf("leader failed to apply: %s", resp.GetError())
+	}
+
+	return nil
+}
+
+// Update Shutdown to clean up RPC resources.
+func (n *Node) Shutdown() error {
+	if n.rpcServer != nil {
+		n.rpcServer.GracefulStop()
+	}
+	if n.rpcClient != nil {
+		n.rpcClient.close()
+	}
+
+	if err := n.raft.Shutdown().Error(); err != nil {
+		return fmt.Errorf("failed to shutdown raft node: %w", err)
+	}
```

Comment:
If any of these raise a `ErrServerClosed` error, it must be ignored. Something like:
```golang
if err := something.Stop(); err != nil && !errors.Is(err, pkg.ErrServerClosed) {
    ...
}
```

Reply:
Addressed in [f6aba9f](https://github.com/gatewayd-io/gatewayd/pull/636/commits/f6aba9f92ff07e7d17787d46a96f15af38f6258b)